ocdata/core/bulk.py

```diff
@@ -0,0 +1,287 @@
+        surface so the bottom is pointing upwards.
+
+        Args:
+            max_miller  An integer indicating the maximum Miller index of the slabs
+                        you are willing to enumerate. Increasing this argument will
+                        increase the number of slabs, but the slabs will
+                        generally become larger.
+        Returns:
+            all_slabs_info  A list of 4-tuples containing:  `pymatgen.Structure`
+                            objects for slabs we have enumerated, the Miller
+                            indices, floats for the shifts, and Booleans for "top".
+        """
+        bulk_struct = standardize_bulk(self.atoms)
+
+        all_slabs_info = []
+        for millers in get_symmetrically_distinct_miller_indices(
+            bulk_struct, MAX_MILLER
+        ):
+            slab_gen = SlabGenerator(
+                initial_structure=bulk_struct,
+                miller_index=millers,
+                min_slab_size=7.0,
+                min_vacuum_size=20.0,
+                lll_reduce=False,
+                center_slab=True,
+                primitive=True,
+                max_normal_search=1,
+            )
+            slabs = slab_gen.get_slabs(
+                tol=0.3, bonds=None, max_broken_bonds=0, symmetrize=False
+            )
+
+            # Additional filtering for the 2D materials' slabs
+            if self.mpid is not None and self.mpid in COVALENT_MATERIALS_MPIDS:
+                slabs = [slab for slab in slabs if is_2D_slab_reasonable(slab) is True]
+
+            # If the bottom of the slabs are different than the tops, then we
+            # want to consider them too.
+            if len(slabs) != 0:
+                flipped_slabs_info = [
+                    (flip_struct(slab), millers, slab.shift, False)
+                    for slab in slabs
+                    if is_structure_invertible(slab) is False
+                ]
+
+                # Concatenate all the results together
+                slabs_info = [(slab, millers, slab.shift, True) for slab in slabs]
+                all_slabs_info.extend(slabs_info + flipped_slabs_info)
+
+        return all_slabs_info
+
+    def __len__(self):
+        return len(self.atoms)
+
+    def __str__(self):
+        return f"Bulk: {self.atoms.get_chemical_formula()}"
+
+    def __repr__(self):
+        return self.__str__()
+
+
+def is_2D_slab_reasonable(struct: pymatgen.core.structure.Structure):
+    """
+    There are 400+ 2D bulk materials whose slabs generated by pymatgen require
+    additional filtering: some slabs are cleaved where one or more surface atoms
+    have no bonds with other atoms on the slab.
+
+    Arguments
+    ---------
+    struct: pymatgen.core.structure.Structure
+        pymatgen structure object of a slab
+
+    Returns
+    -------
+    A boolean indicating whether or not the slab is reasonable, where
+    reasonable is defined as having at least one neighboring atom within 3A.
```

Comment:
I would like to have more info about this. What are we calling a reasonable 2D material slab? I think I would be more rigorous.